### PR TITLE
[Aptos Framework] [Account] Authentication key rotation 

### PR DESCRIPTION
**Pull request by chloeqjz**
### Description
Currently, when a user rotates their authentication key, there is no programmatic relationship between their new authentication key and their account address. If the user’s computer crashes after they rotate the key, they would only be able to restore the most recent authentication key from the seed phrases, but not the original address of the account. This means that the user essentially loses all their assets, as they wouldn't be able to recover their address. We would like to extend our current authentication key rotation library so that we can recover the original address from the user’s most recent authentication key.

This new key rotation function works like the following:
Upon key rotation from old secret key $sk$ to new secret key $sk'$ (with public key $pk'$),

1. We verify that the user owns the new secret key $sk'$
2. We add a new entry `<new authentication key, originating address>` to a global address redirection map, and remove the previous `<old authentication key, originating address>` entry, if there is one

### Test Plan
see e2e and unit tests

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/2972)

### CODE REVIEW

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 1]
```diff
@@ -47,6 +49,15 @@ module aptos_framework::account {
     struct RotationCapability has drop, store { account: address }
     struct SignerCapability has drop, store { account: address }
 
+    struct OriginatingAddress has key, store {
+        address_map: Table<address, address>,
+    }
+
+    struct RotationProof has drop {
+        originator: address, // originating address
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:57]: I'm thinking we should probably add the current sequence number (account.sequence_number) as the first element.

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 2]
```diff
@@ -185,6 +208,40 @@ module aptos_framework::account {
...
+            current_auth_key,
+        };
+
+        assert!(signature::ed25519_verify_t(signature, new_public_key, proof), EINVALID_PROOF_OF_KNOWLEDGE);
+        let address_map = &mut borrow_global_mut<OriginatingAddress>(@aptos_framework).address_map;
+        if (table::contains(address_map, current_auth_key)) {
+            table::remove(address_map, current_auth_key);
+        };
+
+        vector::push_back(&mut new_public_key, 0);
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:238]: let's put a comment on wtf is happening here lol

(ed25519 authenticator is the sha3 256 hash of the public key with a zero appended)

**Review by davidiw — APPROVED** [review]
amazing

File: aptos-move/framework/aptos-stdlib/sources/signature.move [hunk 3]
```diff
@@ -177,6 +184,14 @@ module aptos_std::signature {
         signature: vector<u8>
     ): (vector<u8>, bool);
 
+    public fun ed25519_verify_t<T: drop> (signature: vector<u8>, public_key: vector<u8>, data: T): bool {
+        let encoded = Proof {
+            type_info: type_info::type_of<T>(),
+            inner: data,
+        };
+        ed25519_verify(signature, public_key, bcs::to_bytes(&encoded))
```
Comment (alinush) [inline aptos-move/framework/aptos-stdlib/sources/signature.move:192]: So the wallet code will have to BCS encode this struct when sending a key-rotation transaction.

I assume that's not a problem.

File: aptos-move/framework/aptos-stdlib/sources/signature.move [hunk 4]
```diff
@@ -177,6 +184,14 @@ module aptos_std::signature {
         signature: vector<u8>
     ): (vector<u8>, bool);
 
+    public fun ed25519_verify_t<T: drop> (signature: vector<u8>, public_key: vector<u8>, data: T): bool {
```
Comment (alinush) [inline aptos-move/framework/aptos-stdlib/sources/signature.move:187]: This needs comments. For example,  is this function meant to verify only key rotation proofs?

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 5]
```diff
@@ -185,6 +208,40 @@ module aptos_framework::account {
         account_resource.authentication_key = new_auth_key;
     }
 
+
+    public entry fun rotate_authentication_key_ed25519(account: &signer, signature: vector<u8>, new_public_key: vector<u8>) acquires Account, OriginatingAddress {
```
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:212]: Add /// comments at the top of this function explaining what it does.

Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:212]: Explain what is 'signature': what is the message signed, and under what secret key is it being signed?

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 6]
```diff
@@ -185,6 +208,40 @@ module aptos_framework::account {
...
+        );
+        assert!(
+            vector::length(&signature) == 64,
+            error::invalid_argument(EMALFORMED_PROOF_OF_KNOWLEDGE)
+        );
+
+        let account_resource = borrow_global_mut<Account>(addr);
+        let current_auth_key = create_address(account_resource.authentication_key);
+
+        let proof = RotationProof {
```
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:227]: @davidiw, this struct was supposed to contain the new public key too, no?

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 7]
```diff
@@ -185,6 +208,40 @@ module aptos_framework::account {
...
+            vector::length(&new_public_key) == 32,
+            error::invalid_argument(EMALFORMED_PUBLIC_KEY)
+        );
+        assert!(
+            vector::length(&signature) == 64,
+            error::invalid_argument(EMALFORMED_PROOF_OF_KNOWLEDGE)
+        );
+
+        let account_resource = borrow_global_mut<Account>(addr);
+        let current_auth_key = create_address(account_resource.authentication_key);
```
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:225]: I would expect `create_address` to return an address not an authentication key. I realize they are the same in the current design, but this is confusing. What is going on here?

Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:225]: I guess you are just trying to fetch the current auth key at this address and store it in `proof` which (confusingly) uses `address` as the type of an auth key?

Could we have some basic type safety here?

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 8]
```diff
@@ -496,4 +553,22 @@ module aptos_framework::account {
...
         let _bob = bob;
     }
+
+    #[test(alice = @0xa11ce)]
+    #[expected_failure(abort_code = 65548)]
+    public entry fun test_invalid_public_key(alice: signer) acquires Account, OriginatingAddress {
+        create_account(signer::address_of(&alice));
+        let test_public_key = vector::empty<u8>();
+        let test_signature = vector::empty<u8>();
+        rotate_authentication_key_ed25519(&alice, test_public_key, test_signature);
```
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:563]: Again, folks, **type safety**, please can we have it? By having vector<u8>'s everywhere you can confuse parameters as you've done here: the public key and the signature should be swapped.

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 9]
```diff
@@ -496,4 +553,22 @@ module aptos_framework::account {
...
+    }
+
+    #[test(alice = @0xa11ce)]
+    #[expected_failure(abort_code = 65549)]
+    public entry fun test_invalid_signature(alice: signer) acquires Account, OriginatingAddress {
+        create_account(signer::address_of(&alice));
+        let account_resource = borrow_global_mut<Account>(signer::address_of(&alice));
+        let test_signature = vector::empty<u8>();
+        rotate_authentication_key_ed25519(&alice, account_resource.authentication_key, test_signature);
```
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:572]: Here too: **type safety**, please!

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 2 above)
Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:238]: Uhm.... we don't have a function that computes authentication keys? Can we add one?

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 10]
```diff
@@ -47,6 +49,15 @@ module aptos_framework::account {
     struct RotationCapability has drop, store { account: address }
     struct SignerCapability has drop, store { account: address }
 
+    struct OriginatingAddress has key, store {
+        address_map: Table<address, address>,
+    }
+
+    struct RotationProof has drop {
+        originator: address, // originating address
+        current_auth_key: address, // current auth key
```
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:58]: Add `new_auth_key` field.

File: aptos-move/e2e-move-tests/tests/rotate_auth_key.rs [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+use e2e_move_tests::{assert_success, enable_golden, MoveHarness};
+use move_deps::move_core_types::parser::parse_struct_tag;
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize)]
+struct OriginatingAddress {
+    handle: u128,
+}
+
+#[derive(Serialize, Deserialize)]
```
Comment (alinush) [inline aptos-move/e2e-move-tests/tests/rotate_auth_key.rs:22]: Add comments here please.

File: aptos-move/e2e-move-tests/tests/rotate_auth_key.rs [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+struct Proof {
+    account_address: AccountAddress,
+    module_name: String,
+    struct_name: String,
+    originator: AccountAddress,
```
Comment (alinush) [inline aptos-move/e2e-move-tests/tests/rotate_auth_key.rs:27]: What's the difference between `account_address` and `originator`?

File: aptos-move/e2e-move-tests/tests/rotate_auth_key.rs [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+    transaction::authenticator::AuthenticationKey,
+};
+
+use cached_framework_packages::aptos_stdlib;
+use e2e_move_tests::{assert_success, enable_golden, MoveHarness};
+use move_deps::move_core_types::parser::parse_struct_tag;
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize)]
+struct OriginatingAddress {
```
Comment (alinush) [inline aptos-move/e2e-move-tests/tests/rotate_auth_key.rs:18]: Where is this struct used?

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 7 above)
Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:225]: Actually, I'm confused why isn't this just `current_auth_key = account_resource.authentication_key`?

**Review by alinush — CHANGES_REQUESTED** [review]
1. Can we add a function that computes authentication keys?
2. Document `Proof` struct and `ed25519_verify_t` in `signature.move`
3. Document `rotate_authentication_key_ed25519` in `account.move`
4. Add the new public key (new auth key) as field in the `RotationProof` struct in `account.move`
5. Can we do anything about type safety, or is this part of a bigger change that adds type safety to our Ed25519 Move API (similar to the current [BLS PR](https://github.com/aptos-labs/aptos-core/pull/2980))
6. Comments in `rotate_auth_key.rs` in `e2e-move-tests/`

File: aptos-move/framework/aptos-stdlib/sources/signature.move [hunk 14]
```diff
@@ -9,7 +9,14 @@
 ///     as per the [IETF BLS draft standard](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-bls-signature#section-2.1).
 ///     TODO: Describe APIs
 module aptos_std::signature {
+    use std::bcs;
     use std::option::Option;
+    use aptos_std::type_info::{Self, TypeInfo};
+
+    struct Proof<ProofType> has drop {
```
[the remark this answers is not in the document]
Comment (chloeqjz) [reply aptos-move/framework/aptos-stdlib/sources/signature.move:16]: will add comments!

File: aptos-move/framework/aptos-stdlib/sources/signature.move (hunk 4 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-stdlib/sources/signature.move:187]: for now yes - will add comments soon

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 6 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move:227]: I guess we're passing in the new public key in the function parameter? why do we need it in the struct?

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 7 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move:225]: I think the address and auth key are the same in the current design, except that address is of type `address` and current auth key is of type `vector<u8>`. we wanted to store <address, address> in the OriginatingAddress map so we're doing `create_address` on the auth key here

File: aptos-move/e2e-move-tests/tests/rotate_auth_key.rs (hunk 12 above)
Comment (chloeqjz) [reply aptos-move/e2e-move-tests/tests/rotate_auth_key.rs:27]: `account_address` refers to the account that owns the struct `RotationProof` - which is our `CORE_CODE_ADDRESS` / `aptos_framework` address. 

the originator is the user's first address.

**Comment by davidiw** [comment]
> 1. Can we add a function that computes authentication keys?
We'll do that in a later PR... We need to figure out a way to scale this out.
> 2. Document `Proof` struct and `ed25519_verify_t` in `signature.move`
agreed
> 3. Document `rotate_authentication_key_ed25519` in `account.move`
agreed
> 4. Add the new public key (new auth key) as field in the `RotationProof` struct in `account.move`
the rotation proof is signed by the new private key, so including it seems ... unnecessary?
> 5. Can we do anything about type safety, or is this part of a bigger change that adds type safety to our Ed25519 Move API (similar to the current [BLS PR](https://github.com/aptos-labs/aptos-core/pull/2980))
the latter
> 6. Comments in `rotate_auth_key.rs` in `e2e-move-tests/`
agreed

File: aptos-move/e2e-move-tests/tests/rotate_auth_key.rs (hunk 12 above)
Comment (alinush) [reply aptos-move/e2e-move-tests/tests/rotate_auth_key.rs:27]: Cool! Please add this to the comments!

**Review by alinush — APPROVED** [review]

File: aptos-move/framework/aptos-stdlib/sources/signature.move [hunk 15]
```diff
@@ -9,7 +9,16 @@
 ///     as per the [IETF BLS draft standard](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-bls-signature#section-2.1).
 ///     TODO: Describe APIs
 module aptos_std::signature {
+    use std::bcs;
     use std::option::Option;
+    use aptos_std::type_info::{Self, TypeInfo};
+
+    // Proof a generic type that includes the type info of a Proof
+    // and the corresponding message of ProofType
+    struct Proof<ProofType> has drop {
```
Comment (alinush) [inline aptos-move/framework/aptos-stdlib/sources/signature.move:18]: @davidiw, for now we'll leave this as is, but this `Proof` name should probably be replaced with something more accurate like `Message` or `Challenge`. This is not a proof in any sense of the word: a signature on such a struct would constitute a proof.

I'll rename this once I understand its usage in the next PR.

**Pull request merged**